jinja2: contextfunction has been replaced by pass_context

rschwebel opened the pull request:
Since jinja2 3.0.0, contextfunction has been replaced by pass_function.
Without this patch, a server explodes with

```
flamingo server -s settings.py menu.py --port=8080 --host=localhost
ERROR:flamingo.server:setup failed
  Traceback (most recent call last):
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/utils/imports.py", line 19, in acquire
      item = importlib.import_module(item)
    File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
    File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
    File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
    File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
    File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 790, in exec_module
    File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/templating/__init__.py", line 2, in <module>
      from .jinja2 import Jinja2  # NOQA
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/templating/jinja2.py", line 10, in <module>
      from jinja2 import Environment, FileSystemLoader, contextfunction
  ImportError: cannot import name 'contextfunction' from 'jinja2' ([...]/env-3.9.2/lib/python3.9/site-packages/jinja2/__init__.py)

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/server/server.py", line 132, in setup
      self.build_environment.setup()
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/server/build_environment.py", line 40, in setup
      self.context.setup()
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/context.py", line 47, in setup
      templating_engine_class, path = acquire(
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/utils/imports.py", line 30, in acquire
      module = importlib.import_module(module_name)
    File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
    File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
    File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 790, in exec_module
    File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/templating/__init__.py", line 2, in <module>
      from .jinja2 import Jinja2  # NOQA
    File "[...]/env-3.9.2/lib/python3.9/site-packages/flamingo/core/templating/jinja2.py", line 10, in <module>
      from jinja2 import Environment, FileSystemLoader, contextfunction
  ImportError: cannot import name 'contextfunction' from 'jinja2' ([...]/env-3.9.2/lib/python3.9/site-packages/jinja2/__init__.py)
```

Signed-off-by: Robert Schwebel <r.schwebel@pengutronix.de>

ukleinek commented on flamingo/core/templating/jinja2.py:
```diff
@@ -38,6 +39,14 @@
+if hasattr(jinja2, "pass_context"):
+    pass_context = jinja2.pass_context
+else:
+    pass_context = jinja2.contextfunction
```
To fix the codecov warning, you might do:

    try:
        from jinja2 import contextfunction
    except ImportError
        from jinja2 import pass_context as contextfunction

rschwebel replied:
Another alternative could be to drop jinja2 < 3 support.

rschwebel replied:
In 3.0.x, contextfunction has been deprecated and pass_context does already exist. Starting from 3.1.0, contextfunction has been removed. I'll improve the description.

rschwebel replied:
The try...except variant above has the disadvantage that it uses the deprecated contextversion in jinja2 3.0.x. I think I leave it as it is.

ukleinek replied:
Well, up to now the code uses `contextfunction`, so that seem natural to me. If you want to convert to `pass_context`, just do the construct the other way round. IMHO a try on an import is more pythonic than importing the whole module and doing hasattr on it.

rschwebel replied:
I tried this, but while testing I found out that we get other errors (from markup_safe) when testing with jinja2 < 3.0.0. So I'm abandoning this variant now completely and require >= 3.0.0.

jluebbe replied:
It seems we don't have any need to stay compatible with jinja2 < 3, so I prefer this simpler solution.

rschwebel commented:
I've reworked the patch a bit, to avoid lgtm errors that suggest not to mix "import jinja2" and "from jinja2 import ...". And added a patch on top, to avoid a warning.

rschwebel commented:
With the new patch that changes travis to "focal", it is now also green. So the python 3.7 failures were indeed reated to the old ubuntu version.

ukleinek reviewed (commented):
typo in commit log: s/documetation/documentation/

ukleinek commented:
> With the new patch that changes travis to "focal", it is now also green. So the python 3.7 failures were indeed reated to the old ubuntu version.

If it was green before your code change, I wonder if you understood what made it red and if it's possible to prevent that.

rschwebel commented:
@ukleinek documentation -> fixed, thx. Travis was not green before; IMHO, this is the last travis build on master: https://app.travis-ci.com/github/pengutronix/flamingo/builds/248450467
